File: Tools/CsvExport.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;

using Android.App;
using Android.Content;
using Android.Support.V4.Content;

namespace VorratsUebersicht
{

    //
    // Anhand vom https://github.com/techtribeyt/androidcsv/tree/master
    //
    internal class CsvExport
    {
        private SQLite.SQLiteConnection databaseConnection;
        private Activity context;
        private string trennzeichen = ",";

        public static string ExportArticles(Activity context, int requestCode)
        {
            SQLite.SQLiteConnection databaseConnection = Android_Database.Instance.GetConnection();
            if (databaseConnection == null)
                return null;

            CsvExport export = new CsvExport();
            export.context = context;
            export.databaseConnection = databaseConnection;

            var result = export.GetArticlesAsCsvString();
            return export.WriteToFile("Vue-Artikel.csv", result, requestCode);
        }

        public static string ExportStorageItems(Activity context, int requestCode)
        {
            SQLite.SQLiteConnection databaseConnection = Android_Database.Instance.GetConnection();
            if (databaseConnection == null)
                return null;

            CsvExport export = new CsvExport();
            export.context = context;
            export.databaseConnection = databaseConnection;

            var result = export.GetStorageItemsAsCsvString();
            return export.WriteToFile("Vue-Lagerbestand.csv", result, requestCode);
        }

        private string WriteToFile(string fileName, StringBuilder result, int requestCode)
        {
            string destination = Environment.GetFolderPath (Environment.SpecialFolder.Personal);
            destination = Path.Combine(destination, fileName);

            //saving the file into device
            using(var writer = new StreamWriter(destination, false))
            {
                writer.Write(result.ToString());
            }
                
            Java.IO.File filelocation = new Java.IO.File(destination);
            var path = FileProvider.GetUriForFile(context, "de.stryi.exportcsv.fileprovider", filelocation);
            Intent fileIntent = new Intent(Intent.ActionSend);
            fileIntent.SetType("text/csv");
            fileIntent.PutExtra(Intent.ExtraSubject, fileName);
            fileIntent.SetFlags(ActivityFlags.NewTask);
            fileIntent.AddFlags(ActivityFlags.GrantReadUriPermission);

            fileIntent.PutExtra(Intent.ExtraStream, path);
            context.StartActivityForResult(Intent.CreateChooser(fileIntent, "CSV Datei senden"), requestCode);

            return destination;
        }

        private StringBuilder GetArticlesAsCsvString()
        {
            StringBuilder header = new StringBuilder();
            header.Append("ArticleId|EANCode|Name|Manufacturer|Category|SubCategory|DurableInfinity|WarnInDays|Size|Unit|Notes|MinQuantity|PrefQuantity|StorageName|Supermarket|Calorie|Price");
            header.AppendLine();
            header.Replace("|", this.trennzeichen);

            StringBuilder data = new StringBuilder();
            var articleList = databaseConnection.Query<Article>("SELECT * FROM Article ORDER BY ArticleId");
            foreach(Article article in articleList)
            {
                StringBuilder row = new StringBuilder();
                this.AddField(row, article.ArticleId);
                this.AddField(row, article.EANCode);
                this.AddField(row, article.Name);
                this.AddField(row, article.Manufacturer);
                this.AddField(row, article.Category);
                this.AddField(row, article.SubCategory);
                this.AddField(row, article.DurableInfinity);
                this.AddField(row, article.WarnInDays);
                this.AddField(row, article.Size);
                this.AddField(row, article.Unit);
                this.AddField(row, article.Notes);
                this.AddField(row, article.MinQuantity);
                this.AddField(row, article.PrefQuantity);
                this.AddField(row, article.StorageName);
                this.AddField(row, article.Supermarket);
                this.AddField(row, article.Calorie);
                this.AddField(row, article.Price);

                data.Append(row);
                data.AppendLine();
            }

            header.Append(data);

            return header;
        }

        private StringBuilder GetStorageItemsAsCsvString()
        {
            StringBuilder header = new StringBuilder();
            header.Append("StorageItemId|ArticleId|Name|Manufacturer|ArticleStorageName|DurableInfinity|WarnInDays|Quantity|BestBefore|StorageName");
            header.AppendLine();
            header.Replace("|", this.trennzeichen);

            StringBuilder data = new StringBuilder();
            var articleList = databaseConnection.Query<StorageItemQuantityResult>(
                "SELECT StorageItemId, StorageItem.ArticleId, Name, Manufacturer, Article.StorageName AS ArticleStorageName, DurableInfinity, WarnInDays, Quantity, BestBefore, StorageItem.StorageName" +
                " FROM StorageItem" +
                " JOIN Article ON StorageItem.ArticleId = Article.ArticleId" +
                " ORDER BY StorageItem.ArticleId, BestBefore DESC");

            foreach(StorageItemQuantityResult item in articleList)
            {
                StringBuilder row = new StringBuilder();
                this.AddField(row, item.StorageItemId);
                this.AddField(row, item.ArticleId);
                this.AddField(row, item.Name);
                this.AddField(row, item.Manufacturer);
                this.AddField(row, item.ArticleStorageName);
                this.AddField(row, item.DurableInfinity);
                this.AddField(row, item.WarnInDays);
                this.AddField(row, item.Quantity);
                this.AddField(row, item.BestBefore);
                this.AddField(row, item.StorageName);

                data.Append(row);
                data.AppendLine();
            }

            header.Append(data);

            return header;
        }

        #region private

        private void AddField(StringBuilder data, bool value)
        {
            if (data.Length > 0)
                data.Append(this.trennzeichen);

            data.Append(value);
        }

        private void AddField(StringBuilder data, int value)
        {
            if (data.Length > 0)
                data.Append(this.trennzeichen);

            data.Append(value);
        }
        private void AddField(StringBuilder data, int? value)
        {
            if (data.Length > 0)
                data.Append(this.trennzeichen);

            if (value == null)
                return;

            data.Append(value.Value);
        }

        private void AddField(StringBuilder data, decimal? value)
        {
            if (data.Length > 0)
                data.Append(this.trennzeichen);

            if (value == null)
                return;

            var text = value.Value.ToString(CultureInfo.CurrentCulture);
            if (text.Contains(this.trennzeichen))
            {
                text = string.Format("\"{0}\"", text);
            }
            data.Append(text);
        }

        private void AddField(StringBuilder data, DateTime? value)
        {
            if (data.Length > 0)
                data.Append(this.trennzeichen);

            if (value == null)
                return;

            data.Append(value.Value.ToShortDateString());
        }

        private void AddField(StringBuilder data, string text)
        {
            if (data.Length > 0)
                data.Append(this.trennzeichen);

            if (text == null)
                return;


            if (text.Contains(','))
            {
                if (text.Contains("\""))
                    text = text.Replace("\"", "\\\"");

                data.Append("\"" + text + "\"");
            }
            else
            {
                data.Append(text);
            }
        }

        #endregion
    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/tools/CsvExport.kt
package de.stryi.vorratsuebersicht.tools

import android.app.Activity
import android.content.Intent
import androidx.core.content.FileProvider
import de.stryi.vorratsuebersicht.database.Database
import java.io.File
import java.time.LocalDate

class CsvExport private constructor(
    private val context: Activity,
    private val separator: String)
{

    companion object {

        fun exportArticles(context: Activity, separator: String)
        {
            val exporter = CsvExport(context, separator)
            val csv = exporter.getArticlesAsCsvString()
            exporter.writeToFileAndSend("Vue-Artikel.csv", csv)
        }

        fun exportStorageItems(context: Activity, separator: String)
        {
            val exporter = CsvExport(context, separator)
            val csv = exporter.getStorageItemsAsCsvString()
            exporter.writeToFileAndSend("Vue-Lagerbestand.csv", csv)
        }
    }

    private fun getArticlesAsCsvString(): StringBuilder {
        val header = StringBuilder()
            .append(
                "ArticleId|EANCode|Name|Manufacturer|Category|SubCategory|" +
                        "DurableInfinity|WarnInDays|Size|Unit|Notes|MinQuantity|" +
                        "PrefQuantity|StorageName|Supermarket|Calorie|Price"
            )
            .appendLine()
            .apply { replace(0, length, toString().replace("|", separator)) }

        val data = StringBuilder()
        val articles = Database.getArticleListToCsvExport()

        for (article in articles) {
            val row = StringBuilder()
            addField(row, article.articleId)
            addField(row, article.eanCode)
            addField(row, article.name)
            addField(row, article.manufacturer)
            addField(row, article.category)
            addField(row, article.subCategory)
            addField(row, article.durableInfinity)
            addField(row, article.warnInDays)
            addField(row, article.size)
            addField(row, article.unit)
            addField(row, article.notes)
            addField(row, article.minQuantity)
            addField(row, article.prefQuantity)
            addField(row, article.storageName)
            addField(row, article.supermarket)
            addField(row, article.calorie)
            addField(row, article.price)

            data.append(row).appendLine()
        }

        header.append(data)
        return header
    }

    private fun getStorageItemsAsCsvString(): StringBuilder {
        val header = StringBuilder()
            .append(
                "StorageItemId|ArticleId|Name|Manufacturer|Category|SubCategory|ArticleStorageName|" +
                        "DurableInfinity|WarnInDays|Quantity|BestBefore|StorageName"
            )
            .appendLine()
            .apply { replace(0, length, toString().replace("|", separator)) }

        val data = StringBuilder()
        val items = Database.getStorageItemQuantityListToCsvExport()
        for (item in items) {
            val row = StringBuilder()
            addField(row, item.storageItemId)
            addField(row, item.articleId)
            addField(row, item.name)
            addField(row, item.manufacturer)
            addField(row, item.category)
            addField(row, item.subcategory)
            addField(row, item.articleStorageName)
            addField(row, item.durableInfinity)
            addField(row, item.warnInDays)
            addField(row, item.quantity)
            addField(row, item.bestBefore)
            addField(row, item.storageName)

            data.append(row).appendLine()
        }

        header.append(data)
        return header
    }

    private fun writeToFileAndSend(fileName: String, content: StringBuilder)
    {
        val dir = context.externalCacheDirs.firstOrNull() ?: context.cacheDir
        val file = File(dir, fileName)

        file.writeText(content.toString())

        val uri = FileProvider.getUriForFile(
            context,
            "de.stryi.vorratsuebersicht.provider",
            file
        )

        val intent = Intent(Intent.ACTION_SEND)
        intent.type = "text/csv"
        intent.putExtra(Intent.EXTRA_SUBJECT, fileName)
        intent.putExtra(Intent.EXTRA_STREAM, uri)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)

        context.startActivity(
            Intent.createChooser(intent, "CSV Datei senden"))
    }

    /* ---------- Helper ---------- */

    private fun appendSeparator(sb: StringBuilder) {
        if (sb.isNotEmpty()) sb.append(separator)
    }

    private fun addField(sb: StringBuilder, value: Boolean) {
        appendSeparator(sb)
        sb.append(value)
    }

    private fun addField(sb: StringBuilder, value: Int) {
        appendSeparator(sb)
        sb.append(value)
    }

    private fun addField(sb: StringBuilder, value: Int?) {
        appendSeparator(sb)
        sb.append(Tools.formatNumber(value))
    }

    private fun addField(sb: StringBuilder, value: Double?) {
        appendSeparator(sb)
        value ?: return

        var text = Tools.formatNumber(value)
        if (text.contains(separator)) {
            text = "\"$text\""
        }
        sb.append(text)
    }

    private fun addField(sb: StringBuilder, value: LocalDate?) {
        appendSeparator(sb)
        if (value == null) {
            return
        }

        sb.append(Tools.toHumanString(value))
    }

    private fun addField(sb: StringBuilder, text: String?) {
        appendSeparator(sb)
        text ?: return

        if (text.contains(separator)) {
            val escaped = text.replace("\"", "\\\"")
            sb.append("\"").append(escaped).append("\"")
        } else {
            sb.append(text)
        }
    }
}